www/source/blog/2017-10-27-distributed-java-application.html.md

```diff
@@ -0,0 +1,360 @@
+
+### Service Bindings
+
+Bindings, expressed by `pkg_binds` and `pkg_exports` facilitate interaction between multiple services. Lets face it, most applications are comprised of more than a single service. When we think of our core Java application, we may think of the Tomcat process running on a web server. However if that was all Habitat ran, we'd likely all walk away disappointed. Our particular Java application shows all of the national parks in the United States. However, without a database service, there would be no parks to display. Furthermore, because our application will be very popular (Right?!) we will need a load balancer to fan traffic out accross multiple web servers. So our application will need to "find" other services that provide a database and our load balancer service will need to find all of the instances running our application to load into its configuration.
+
+`pkg_binds` and `pkg_exports` constitute two sides of a contract between services. `pkg_binds` declares the **consumer** side of this contract and specifies what configuration properties a consuming service expects a **provider** service to expose. Our plan above states that it requires a `database` service and that service must expose a `port` property. If we were to run our application in a Habitat Supervisor and a service could not be found that fulfills the contract, our service would not start until our database service shows up.
+
+`pkg_exports` declares what configuration we want to expose as public properties to external services. Here we want to expose a property called `port` and we will map our internal `tomcat_port` property to that external `port` alias. In our distributed Java application, the `haproxy` service is going to need this. In fact, if were were to look at the [`haproxy` plan](https://github.com/habitat-sh/core-plans/blob/master/haproxy/plan.sh), we would find:
```

Comment:
It might be worth mentioning where the `port` and `tomcat_port` values come from. Are they defined in some TOML somewhere? Since this is a shell script, are they variable references or functions? When/how do they receive their values?

Reply:
... or if we'll learn about that later, it'd be fine to say that too, either way.